### PR TITLE
Merge std and core libraries

### LINKED ISSUE #6708: Combine std-lib and core
The std-lib and core should be combined into a single library

### PR DESCRIPTION
**Pull request by SwayStar123**
## Description
Merges the two libraries. They were initially separate to separate the core logic and fuel vm specific functionality, but that separation is no longer maintained so having a merged library is better.

Closes #6708

## Checklist

- [ ] I have linked to any relevant issues.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
   - [ ] If my change requires substantial documentation changes, I have [requested support from the DevRel team](https://github.com/FuelLabs/devrel-requests/issues/new/choose)
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [ ] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [ ] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: forc-pkg/src/pkg.rs [hunk 1]
```diff
@@ -1640,59 +1638,11 @@ pub fn dependency_namespace(
...
         };
         root_namespace.add_external(dep_name, dep_namespace);
-        let dep = &graph[dep_node];
-        if dep.name == CORE {
-            core_added = true;
-        }
-    }
-
-    // Add `core` if not already added.
-    if !core_added {
```
Comment (xunilrj) [inline forc-pkg/src/pkg.rs:1650]: Shouldn´t this method do the same thing, but for the `std` lib?

File: forc-plugins/forc-doc/src/tests/data/impl_traits_generic/Forc.toml [hunk 2]
```diff
@@ -3,6 +3,7 @@ authors = ["Fuel Labs <contact@fuel.sh>"]
 entry = "lib.sw"
 license = "Apache-2.0"
 name = "impl_traits_generic"
+implicit-std = false
 
 [dependencies]
-core = { path = "../../../../../../sway-lib-core" }
+ops = { path = "../ops" }
```
Comment (xunilrj) [inline forc-plugins/forc-doc/src/tests/data/impl_traits_generic/Forc.toml:9]: Why `ops` here and not `std`?

Comment (SwayStar123) [reply forc-plugins/forc-doc/src/tests/data/impl_traits_generic/Forc.toml:9]: previously the test used core as it wasnt that large, so the hardcoded values in the test werent that long. but with the new std its comically long and unnecessary.

File: forc-pkg/src/pkg.rs (hunk 1 above)
Comment (SwayStar123) [reply forc-pkg/src/pkg.rs:1650]: does the implicitly_include_std_if_missing function not do it?

File: sway-lib-std/src/ops.sw [hunk 3]
```diff
@@ -1743,36 +1744,124 @@ impl PartialEq for str {
...
+    pub fn is_zero(self) -> bool {
+        self == 0u8
     }
 }
 
-#[test]
-pub fn ok_str_eq() {
-    assert("" == "");
-    assert("a" == "a");
+impl u16 {
```
Comment (xunilrj) [inline sway-lib-std/src/ops.sw:1767]: If we are moving this, why to `ops.sw` and not to `primitives.sw`? I imagine it's the ordering of these modules.

Comment (SwayStar123) [reply sway-lib-std/src/ops.sw:1767]: the implementation uses the eq operation

**Review by IGI-111 — CHANGES_REQUESTED** [review]
I'm going through the PR and it looks fine so far but I want to block merging this until we have the migrations for it merged and released.

File: test/src/e2e_vm_tests/test_programs/should_pass/language/nested_while_and_if/src/main.sw [hunk 4]
```diff
@@ -1,6 +1,6 @@
 script;
 
-use core::*;
+use std::*;
```
Comment (jjcnn) [inline test/src/e2e_vm_tests/test_programs/should_pass/language/nested_while_and_if/src/main.sw:3]: This line of code doesn't actually do anything, so we should be able to remove it.

File: test/src/e2e_vm_tests/test_programs/should_pass/language/new_allocator_test/src/main.sw [hunk 5]
```diff
@@ -1,6 +1,6 @@
 script;
 
-use core::*;
+use std::*;
```
Comment (jjcnn) [inline test/src/e2e_vm_tests/test_programs/should_pass/language/new_allocator_test/src/main.sw:3]: Same issue here: The line doesn't do anything, and can be removed.

File: test/src/e2e_vm_tests/test_programs/should_pass/require_contract_deployment/bal_opcode/Forc.toml [hunk 6]
```diff
@@ -6,4 +6,4 @@ name = "bal_opcode"
 
 [dependencies]
 balance_test_abi = { path = "../../test_abis/balance_test_abi" }
-std = { path = "../../../../../../../sway-lib-std" }
+std = { path = "../../../../../../../sway-lib-std" }
```
Comment (jjcnn) [inline test/src/e2e_vm_tests/test_programs/should_pass/require_contract_deployment/bal_opcode/Forc.toml:9]: Only whitespace changes in this file.

File: test/src/e2e_vm_tests/test_programs/should_pass/require_contract_deployment/caller_auth_test/Forc.toml [hunk 7]
```diff
@@ -6,4 +6,5 @@ name = "caller_auth_test"
 
 [dependencies]
 auth_testing_abi = { path = "../../test_abis/auth_testing_abi" }
-std = { path = "../../../../reduced_std_libs/sway-lib-std-assert" }
+std = { path = "../../../../../../../sway-lib-std" }
```
Comment (jjcnn) [inline test/src/e2e_vm_tests/test_programs/should_pass/require_contract_deployment/caller_auth_test/Forc.toml:9]: If `sway-lib-std-assert` doesn't contain everything this test needs, then maybe there's a case for creating a new reduced std library containing stuff related to contract calls and deployment?

Probably best to do that as a separate issue.

File: test/src/e2e_vm_tests/test_programs/should_pass/test_abis/abi_with_tuples/Forc.toml [hunk 8]
```diff
@@ -5,4 +5,4 @@ license = "Apache-2.0"
 name = "abi_with_tuples"
 
 [dependencies]
-core = { path = "../../../../../../../sway-lib-core" }
\ No newline at end of file
+std = { path = "../../../../reduced_std_libs/sway-lib-std-assert" }
```
Comment (jjcnn) [inline test/src/e2e_vm_tests/test_programs/should_pass/test_abis/abi_with_tuples/Forc.toml:8]: I think this test only needs sway-lib-std-core, not sway-lib-std-assert.

File: test/src/e2e_vm_tests/test_programs/should_pass/test_abis/array_of_structs_abi/Forc.toml [hunk 9]
```diff
@@ -5,4 +5,4 @@ license = "Apache-2.0"
 name = "array_of_structs_abi"
 
 [dependencies]
-core = { path = "../../../../../../../sway-lib-core" }
\ No newline at end of file
+std = { path = "../../../../../../../sway-lib-std" }
```
Comment (jjcnn) [inline test/src/e2e_vm_tests/test_programs/should_pass/test_abis/array_of_structs_abi/Forc.toml:8]: I think this test only needs sway-lib-std-core, not the full std.

File: test/src/e2e_vm_tests/test_programs/should_pass/test_abis/auth_testing_abi/Forc.toml [hunk 10]
```diff
@@ -5,4 +5,4 @@ license = "Apache-2.0"
 name = "auth_testing_abi"
 
 [dependencies]
-core = { path = "../../../../../../../sway-lib-core" }
+std = { path = "../../../../../../../sway-lib-std" }
```
Comment (jjcnn) [inline test/src/e2e_vm_tests/test_programs/should_pass/test_abis/auth_testing_abi/Forc.toml:8]: I think this test only needs sway-lib-std-core, not the full std.

File: test/src/e2e_vm_tests/test_programs/should_pass/test_abis/balance_test_abi/Forc.toml [hunk 11]
```diff
@@ -5,4 +5,4 @@ license = "Apache-2.0"
 name = "balance_test_abi"
 
 [dependencies]
-core = { path = "../../../../../../../sway-lib-core" }
\ No newline at end of file
+std = { path = "../../../../../../../sway-lib-std" }
```
Comment (jjcnn) [inline test/src/e2e_vm_tests/test_programs/should_pass/test_abis/balance_test_abi/Forc.toml:8]: I think this test only needs sway-lib-std-core, not the full std.

File: test/src/e2e_vm_tests/test_programs/should_pass/test_abis/storage_access_abi/Forc.toml [hunk 12]
```diff
@@ -5,4 +5,4 @@ license = "Apache-2.0"
 name = "storage_access_abi"
 
 [dependencies]
-core = { path = "../../../../../../../sway-lib-core" }
+std = { path = "../../../../../../../sway-lib-std" }
```
Comment (jjcnn) [inline test/src/e2e_vm_tests/test_programs/should_pass/test_abis/storage_access_abi/Forc.toml:8]: I think this test only needs sway-lib-std-core, not the full std.

File: test/src/e2e_vm_tests/test_programs/should_pass/test_abis/storage_enum_abi/Forc.toml [hunk 13]
```diff
@@ -5,4 +5,4 @@ license = "Apache-2.0"
 name = "storage_enum_abi"
 
 [dependencies]
-core = { path = "../../../../../../../sway-lib-core" }
+std = { path = "../../../../../../../sway-lib-std" }
```
Comment (jjcnn) [inline test/src/e2e_vm_tests/test_programs/should_pass/test_abis/storage_enum_abi/Forc.toml:8]: I think this test only needs sway-lib-std-core, not the full std.

File: test/src/e2e_vm_tests/test_programs/should_pass/test_contracts/abi_with_tuples_contract/Forc.toml [hunk 14]
```diff
@@ -6,5 +6,5 @@ name = "abi_with_tuples_contract"
 implicit-std = false
 
 [dependencies]
-core = { path = "../../../../../../../sway-lib-core" }
+std = { path = "../../../../reduced_std_libs/sway-lib-std-assert" }
```
Comment (jjcnn) [inline test/src/e2e_vm_tests/test_programs/should_pass/test_contracts/abi_with_tuples_contract/Forc.toml:9]: I think this test only needs sway-lib-std-core.

File: test/src/e2e_vm_tests/test_programs/should_pass/test_contracts/array_of_structs_contract/Forc.toml [hunk 15]
```diff
@@ -7,3 +7,4 @@ implicit-std = false
 
 [dependencies]
 array_of_structs_abi = { path = "../../test_abis/array_of_structs_abi" }
+std = { path = "../../../../../../../sway-lib-std" }
```
Comment (jjcnn) [inline test/src/e2e_vm_tests/test_programs/should_pass/test_contracts/array_of_structs_contract/Forc.toml:10]: Probably only needs sway-lib-std-core.

File: test/src/e2e_vm_tests/test_programs/should_pass/test_contracts/nested_struct_args_contract/Forc.toml [hunk 16]
```diff
@@ -5,5 +5,5 @@ license = "Apache-2.0"
 name = "nested_struct_args_contract"
 
 [dependencies]
-core = { path = "../../../../../../../sway-lib-core" }
+std = { path = "../../../../reduced_std_libs/sway-lib-std-assert" }
```
Comment (jjcnn) [inline test/src/e2e_vm_tests/test_programs/should_pass/test_contracts/nested_struct_args_contract/Forc.toml:8]: Probably only needs sway-lib-std-core.

File: test/src/sdk-harness/test_projects/string_slice/string_slice_predicate/Forc.toml [hunk 17]
```diff
@@ -5,4 +5,4 @@ license = "Apache-2.0"
 name = "string_slice_predicate"
 
 [dependencies]
-core = { path = "../../../../../../sway-lib-core" }
+std = { path = "../../../../../../sway-lib-std" }
```
Comment (jjcnn) [inline test/src/sdk-harness/test_projects/string_slice/string_slice_predicate/Forc.toml:8]: Probably only needs sway-lib-std-core.

**Review by jjcnn — CHANGES_REQUESTED** [review]
Truly a Herculian effort on this one!

I've added a few comments. There shouldn't be much work in them, so there should be time to do them before the release.

The ones about using the reduced std libs can be left for a separate issue if needed.

File: sway-core/src/semantic_analysis/ast_node/declaration/auto_impl/abi_encoding.rs [hunk 18]
```diff
@@ -195,7 +195,12 @@ where
         engines: &Engines,
         decl: &TyDecl,
     ) -> Option<(Option<TyAstNode>, Option<TyAstNode>)> {
-        if self.ctx.namespace.current_package_name().as_str() == "core" {
+        if self.ctx.namespace.current_package_name().as_str() == "std"
+            && matches!(
+                self.ctx.namespace.current_module().name().as_str(),
+                "codec" | "raw_slice" | "raw_ptr" | "ops" | "primitives" | "registers" | "flags"
+            )
+        {
```
[the remark this answers is not in the document]
Comment (SwayStar123) [reply sway-core/src/semantic_analysis/ast_node/declaration/auto_impl/abi_encoding.rs:203]: This isnt so much about them being former members of core, but that they are the dependencies of the codec library. I will add a comment

**Review by jjcnn — APPROVED** [review]
Changing my review status to approved. 

The one change that was definitely needed was comments in abi_encoding.rs, which have been added.

My remaining review comments are nits.

**Review by IGI-111 — APPROVED** [review]

**Review by bitzoic — APPROVED** [review]
I would have liked to see more edge cases covered(zero and max values) in the inline tests but these can be added afterward
